File: sdk/typescript/packages/muzen-sdk/src/wire-validation.ts
import type {
  ModelProfile,
  MuzenWorkerRun,
  ProviderProfile,
  ReviewArtifact,
  ReviewArtifactExport,
  ReviewEvent,
  ReviewResult,
  ReviewSessionSnapshot,
  ReviewSource,
  ReviewStatus,
} from "./types.js";

export function unwrapReviewSnapshot(value: unknown): ReviewSessionSnapshot {
  const snapshot = isRecord(value) && isRecord(value.review) ? value.review : value;
  if (!isReviewSessionSnapshot(snapshot)) {
    throw new Error("Muzen remote returned an invalid review session snapshot");
  }
  return snapshot;
}

export function unwrapWorkerRun(value: unknown): MuzenWorkerRun {
  if (!isRecord(value)) {
    throw new Error("invalid worker run response from muzen-runner");
  }
  const run = value as Record<string, unknown>;
  if (
    typeof run.workerId !== "string" ||
    typeof run.claimed !== "number" ||
    typeof run.completed !== "number" ||
    typeof run.retried !== "number" ||
    typeof run.failed !== "number" ||
    typeof run.skipped !== "number"
  ) {
    throw new Error("invalid worker run response from muzen-runner");
  }
  return {
    workerId: run.workerId,
    claimed: run.claimed,
    completed: run.completed,
    retried: run.retried,
    failed: run.failed,
    skipped: run.skipped,
  };
}

export function unwrapOptionalReviewResult(value: unknown): ReviewResult | undefined {
  if (value === undefined || value === null) {
    return undefined;
  }
  const result = isRecord(value) && "result" in value ? value.result : value;
  if (result === undefined || result === null) {
    return undefined;
  }
  if (!isReviewResult(result)) {
    throw new Error("Muzen remote returned an invalid review result");
  }
  return result;
}

export function unwrapReviewEvents(value: unknown): ReviewEvent[] {
  const events = isRecord(value) && Array.isArray(value.events) ? value.events : value;
  if (!Array.isArray(events) || !events.every(isReviewEvent)) {
    throw new Error("Muzen remote returned invalid review events");
  }
  return events;
}

export function unwrapReviewArtifact(value: unknown): ReviewArtifact {
  const artifact = isRecord(value) && isRecord(value.artifact) ? value.artifact : value;
  if (!isReviewArtifact(artifact)) {
    throw new Error("Muzen remote returned an invalid review artifact");
  }
  return artifact;
}

export function unwrapReviewArtifactExport(value: unknown): ReviewArtifactExport {
  if (!isReviewArtifactExport(value)) {
    throw new Error("Muzen remote returned an invalid artifact export");
  }
  return value;
}

export function unwrapModelProfile(value: unknown): ModelProfile {
  const profile = isRecord(value) && isRecord(value.profile) ? value.profile : value;
  if (!isModelProfile(profile)) {
    throw new Error("Muzen remote returned an invalid model profile");
  }
  return profile;
}

export function unwrapModelProfiles(value: unknown): ModelProfile[] {
  const profiles = isRecord(value) && Array.isArray(value.profiles) ? value.profiles : value;
  if (!Array.isArray(profiles) || !profiles.every(isModelProfile)) {
    throw new Error("Muzen remote returned invalid model profiles");
  }
  return profiles;
}

export function unwrapProviderProfile(value: unknown): ProviderProfile {
  const profile = isRecord(value) && isRecord(value.profile) ? value.profile : value;
  if (!isProviderProfile(profile)) {
    throw new Error("Muzen remote returned an invalid provider profile");
  }
  return profile;
}

export function unwrapProviderProfiles(value: unknown): ProviderProfile[] {
  const profiles = isRecord(value) && Array.isArray(value.profiles) ? value.profiles : value;
  if (!Array.isArray(profiles) || !profiles.every(isProviderProfile)) {
    throw new Error("Muzen remote returned invalid provider profiles");
  }
  return profiles;
}

export function unwrapWebhookHttpResponse(value: unknown): {
  statusCode: number;
  headers: Record<string, string>;
  body: string;
} {
  if (
    isRecord(value) &&
    typeof value.statusCode === "number" &&
    isRecord(value.headers) &&
    typeof value.body === "string"
  ) {
    return {
      statusCode: value.statusCode,
      headers: stringRecord(value.headers),
      body: value.body,
    };
  }
  throw new Error("muzen-runner returned an invalid webhook response");
}

function stringRecord(value: Record<string, unknown>): Record<string, string> {
  const result: Record<string, string> = {};
  for (const [key, item] of Object.entries(value)) {
    if (typeof item === "string") {
      result[key] = item;
    }
  }
  return result;
}

function isReviewSessionSnapshot(value: unknown): value is ReviewSessionSnapshot {
  return (
    isRecord(value) &&
    typeof value.id === "string" &&
    isReviewStatus(value.status) &&
    isReviewSource(value.source) &&
    (value.result === undefined || isReviewResult(value.result))
  );
}

function isReviewSource(value: unknown): value is ReviewSource {
  return (
    isRecord(value) &&
    (value.type === "local" ||
      value.type === "github_pull_request" ||
      value.type === "gitlab_merge_request")
  );
}

function isReviewResult(value: unknown): value is ReviewResult {
  return (
    isRecord(value) &&
    typeof value.reviewId === "string" &&
    typeof value.sessionId === "string" &&
    isReviewStatus(value.status) &&
    (value.conclusion === "approved" ||
      value.conclusion === "commented" ||
      value.conclusion === "changes_requested") &&
    typeof value.summary === "string" &&
    Array.isArray(value.findings) &&
    isRecord(value.coverage)
  );
}

function isReviewEvent(value: unknown): value is ReviewEvent {
  return (
    isRecord(value) &&
    typeof value.cursor === "string" &&
    typeof value.type === "string" &&
    typeof value.reviewId === "string" &&
    typeof value.timestampUtc === "string"
  );
}

function isReviewArtifact(value: unknown): value is ReviewArtifact {
  return (
    isRecord(value) &&
    typeof value.artifactId === "string" &&
    typeof value.bytes === "number" &&
    typeof value.contentHash === "string" &&
    typeof value.content === "string"
  );
}

function isReviewArtifactExport(value: unknown): value is ReviewArtifactExport {
  return (
    isRecord(value) &&
    (value.view === "redacted" || value.view === "raw") &&
    typeof value.artifactCount === "number" &&
    typeof value.totalBytes === "number" &&
    Array.isArray(value.artifacts) &&
    value.artifacts.every(isReviewArtifact)
  );
}

function isReviewStatus(value: unknown): value is ReviewStatus {
  return (
    value === "created" ||
    value === "queued" ||
    value === "running" ||
    value === "completed" ||
    value === "failed" ||
    value === "cancelled"
  );
}

function isModelProfile(value: unknown): value is ModelProfile {
  return (
    isRecord(value) &&
    typeof value.workspaceId === "string" &&
    typeof value.name === "string" &&
    typeof value.version === "string" &&
    isModelProviderKind(value.provider) &&
    typeof value.model === "string" &&
    typeof value.updatedAtUtc === "string"
  );
}

function isProviderProfile(value: unknown): value is ProviderProfile {
  return (
    isRecord(value) &&
    typeof value.workspaceId === "string" &&
    typeof value.name === "string" &&
    typeof value.version === "string" &&
    isSourceProviderKind(value.provider) &&
    typeof value.updatedAtUtc === "string"
  );
}

function isModelProviderKind(value: unknown): value is ModelProfile["provider"] {
  return (
    value === "openai" ||
    value === "anthropic" ||
    value === "openai_compatible"
  );
}

function isSourceProviderKind(value: unknown): value is ProviderProfile["provider"] {
  return value === "github" || value === "gitlab";
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

File: sdk/python/muzen/wire_validation.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

from .runner_mapping import _map_runner_artifact, _map_runner_status
from .types import (
    ModelProfile,
    ProviderProfile,
    ReviewArtifact,
    ReviewArtifactExport,
    ReviewCoverage,
    ReviewEvent,
    ReviewFinding,
    ReviewFindingEvidence,
    ReviewResult,
    ReviewSessionSnapshot,
    ReviewSource,
)


def _unwrap_review_snapshot(value: Any) -> ReviewSessionSnapshot:
    snapshot = value.get("review") if isinstance(value, dict) and isinstance(value.get("review"), dict) else value
    if not isinstance(snapshot, dict):
        raise RuntimeError("Muzen remote returned an invalid review session snapshot")
    return ReviewSessionSnapshot(
        id=snapshot["id"],
        status=_map_runner_status(snapshot["status"]),
        source=_remote_source(snapshot["source"]),
        result=_unwrap_optional_review_result(snapshot.get("result")),
    )


def _unwrap_optional_review_result(value: Any) -> Optional[ReviewResult]:
    if value is None:
        return None
    result = value.get("result") if isinstance(value, dict) and "result" in value else value
    if result is None:
        return None
    if not isinstance(result, dict):
        raise RuntimeError("Muzen remote returned an invalid review result")
    return _remote_result(result)


def _unwrap_review_events(value: Any) -> List[ReviewEvent]:
    events = value.get("events") if isinstance(value, dict) and isinstance(value.get("events"), list) else value
    if not isinstance(events, list):
        raise RuntimeError("Muzen remote returned invalid review events")
    return [
        ReviewEvent(
            cursor=str(event["cursor"]),
            type=event["type"],
            review_id=event["reviewId"],
            timestamp_utc=event["timestampUtc"],
            payload=event.get("payload"),
        )
        for event in events
    ]


def _unwrap_review_artifact(value: Any) -> ReviewArtifact:
    artifact = value.get("artifact") if isinstance(value, dict) and isinstance(value.get("artifact"), dict) else value
    if not isinstance(artifact, dict):
        raise RuntimeError("Muzen remote returned an invalid review artifact")
    return _map_runner_artifact(artifact)


def _unwrap_artifact_export(value: Any) -> ReviewArtifactExport:
    if not isinstance(value, dict):
        raise RuntimeError("Muzen remote returned an invalid artifact export")
    return ReviewArtifactExport(
        view=value.get("view", "redacted"),
        artifact_count=value.get("artifactCount", 0),
        total_bytes=value.get("totalBytes", 0),
        artifacts=[
            _map_runner_artifact(artifact)
            for artifact in value.get("artifacts", [])
            if isinstance(artifact, dict)
        ],
    )


def _unwrap_model_profile(value: Any) -> ModelProfile:
    profile = value.get("profile") if isinstance(value, dict) and isinstance(value.get("profile"), dict) else value
    if not isinstance(profile, dict):
        raise RuntimeError("Muzen remote returned an invalid model profile")
    return ModelProfile(
        workspace_id=profile["workspaceId"],
        name=profile["name"],
        version=profile["version"],
        provider=profile["provider"],
        model=profile["model"],
        secret_ref=profile.get("secretRef"),
        base_url=profile.get("baseUrl"),
        routing=profile.get("routing") or {},
        updated_at_utc=profile.get("updatedAtUtc", ""),
    )


def _unwrap_model_profiles(value: Any) -> List[ModelProfile]:
    profiles = value.get("profiles") if isinstance(value, dict) and isinstance(value.get("profiles"), list) else value
    if not isinstance(profiles, list):
        raise RuntimeError("Muzen remote returned invalid model profiles")
    return [_unwrap_model_profile(profile) for profile in profiles]


def _unwrap_provider_profile(value: Any) -> ProviderProfile:
    profile = value.get("profile") if isinstance(value, dict) and isinstance(value.get("profile"), dict) else value
    if not isinstance(profile, dict):
        raise RuntimeError("Muzen remote returned an invalid provider profile")
    return ProviderProfile(
        workspace_id=profile["workspaceId"],
        name=profile["name"],
        version=profile["version"],
        provider=profile["provider"],
        secret_ref=profile.get("secretRef"),
        base_url=profile.get("baseUrl"),
        routing=profile.get("routing") or {},
        updated_at_utc=profile.get("updatedAtUtc", ""),
    )


def _unwrap_provider_profiles(value: Any) -> List[ProviderProfile]:
    profiles = value.get("profiles") if isinstance(value, dict) and isinstance(value.get("profiles"), list) else value
    if not isinstance(profiles, list):
        raise RuntimeError("Muzen remote returned invalid provider profiles")
    return [_unwrap_provider_profile(profile) for profile in profiles]


def _remote_source(value: Dict[str, Any]) -> ReviewSource:
    if value["type"] == "local":
        return ReviewSource(
            type="local",
            repo=value["repo"],
        )
    if value["type"] == "raw_snapshot":
        return ReviewSource(
            type="raw_snapshot",
            root=value["root"],
        )
    if value["type"] == "perforce_changelist":
        return ReviewSource(
            type="perforce_changelist",
            server=value.get("server"),
            changelist=value.get("changelist"),
            client=value.get("client"),
            depot_paths=value.get("depotPaths", []),
        )
    if value["type"] == "custom":
        return ReviewSource(
            type="custom",
            provider=value.get("provider"),
            id=value.get("id"),
        )
    return ReviewSource(
        type=value["type"],
        owner=value.get("owner"),
        repo=value.get("repo"),
        number=value.get("number"),
    )


def _remote_result(value: Dict[str, Any]) -> ReviewResult:
    return ReviewResult(
        review_id=value["reviewId"],
        session_id=value["sessionId"],
        status=_map_runner_status(value["status"]),
        conclusion=value["conclusion"],
        summary=value["summary"],
        findings=[
            ReviewFinding(
                id=finding.get("id", ""),
                severity=finding.get("severity", "info"),
                category=finding.get("category", "other"),
                title=finding.get("title", ""),
                message=finding.get("message", ""),
                location=finding.get("location"),
                suggested_fix=finding.get("suggestedFix"),
                confidence=finding.get("confidence"),
                validation_status=finding.get("validationStatus"),
                evidence=[
                    ReviewFindingEvidence(
                        evidence_id=evidence.get("evidenceId", ""),
                        artifact_id=evidence.get("artifactId", ""),
                        kind=evidence.get("kind", ""),
                        content_hash=evidence.get("contentHash", ""),
                        producing_tool_call_id=evidence.get("producingToolCallId", ""),
                    )
                    for evidence in finding.get("evidence", [])
                    if isinstance(evidence, dict)
                ],
                discovered_by=list(finding.get("discoveredBy", [])),
                validated_by=list(finding.get("validatedBy", [])),
                challenged_by=list(finding.get("challengedBy", [])),
            )
            for finding in value.get("findings", [])
        ],
        coverage=ReviewCoverage(
            files_considered=value.get("coverage", {}).get("filesConsidered", 0),
            files_reviewed=value.get("coverage", {}).get("filesReviewed", 0),
            files_skipped=value.get("coverage", {}).get("filesSkipped", 0),
        ),
        metadata=value.get("metadata") or {},
    )
